framework/sandstone_ssl_rand.cpp

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2023 Intel Corporation.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#include "sandstone_ssl.h"
+#include "sandstone_ssl_rand.h"
+
+static OSSL_PROVIDER *s_prov;
+
+/* Random context management */
+static void *sandstone_rand_newctx(void *ctx, void *parent,
+                            const OSSL_DISPATCH *parent_calls)
+{
+    // Allocate context
+    int *rand_st = reinterpret_cast<int*>(malloc(sizeof(*rand_st)));
+    if (rand_st != NULL)
+        *rand_st = EVP_RAND_STATE_UNINITIALISED;
+
+    return rand_st;
+}
+
+static void sandstone_rand_freectx(void *ctx)
+{
+     free(ctx);
+}
+
+/* Random number generator functions */
+static int sandstone_rand_instantiate(void *ctx, unsigned int strength, int prediction_resistance,
+                               const unsigned char *pstr, size_t pstr_len,
+                               const OSSL_PARAM params[])
+{
+    ctx = reinterpret_cast<void *>(EVP_RAND_STATE_READY);
+    return 1;
+}
+
+static int sandstone_rand_uninstantiate(void *ctx)
+{
+    ctx = reinterpret_cast<void *>(EVP_RAND_STATE_UNINITIALISED);
+    return 1;
+}
+
+static int sandstone_rand_generate(void *ctx, unsigned char *out, size_t outlen,
+                            unsigned int strength, int prediction_resistance,
+                            const unsigned char *addin, size_t addin_len)
+{
+    memset_random(out, outlen);
+
+    return 1;
+}
+
+// TODO: Do something
+static int sandstone_rand_reseed(void *ctx, int prediction_resistance, const unsigned char *ent,
+                          size_t ent_len, const unsigned char *addin, size_t addin_len)
+{
+    return 1;
+}
+
+static int sandstone_rand_enable_locking(ossl_unused void *ctx)
+{
+    return 1;
+}
+
+// TODO: Fix macros support and add prefix
+static const OSSL_PARAM *sandstone_rand_gettable_ctx_params(void *ctx, void *provctx)
+{
+    static const OSSL_PARAM known_gettable_ctx_params[] = {
+        OSSL_PARAM_int(OSSL_RAND_PARAM_STATE, NULL),
+        OSSL_PARAM_uint(OSSL_RAND_PARAM_STRENGTH, NULL),
+        OSSL_PARAM_size_t(OSSL_RAND_PARAM_MAX_REQUEST, NULL),
+        OSSL_PARAM_END
+    };
+    return known_gettable_ctx_params;
+}
+
+/* Get details of currently set parameter values associated with the given provider
+ * https://www.openssl.org/docs/man3.0/man7/EVP_RAND-CTR-DRBG.html
+ */
+static int sandstone_rand_get_ctx_params(void *ctx, OSSL_PARAM params[])
+{
+    OSSL_PARAM *p;
+
+    p = s_OSSL_PARAM_locate(params, OSSL_RAND_PARAM_STATE);
+    if (p != NULL && !s_OSSL_PARAM_set_int(p, *(int *)ctx))
+        return 0;
+
+    p = s_OSSL_PARAM_locate(params, OSSL_RAND_PARAM_STRENGTH);
+    if (p != NULL && !s_OSSL_PARAM_set_int(p, 500))
+        return 0;
+
+    p = s_OSSL_PARAM_locate(params, OSSL_RAND_PARAM_MAX_REQUEST);
```

Comment:
What's the max request? Is that data length or the maximum value we return?

Reply:
Good question, from documentation: "Specifies the maximum number of bytes that can be generated in a single call to OSSL_FUNC_rand_generate".  Do we have a limit for memset_random?

Reply:
We don't.